src/metabase/driver/sql/query_processor.clj

```diff
@@ -673,6 +679,24 @@
   [driver [_ arg amount unit]]
   (add-interval-honeysql-form driver (->honeysql driver arg) (- amount) unit))
 
+(defn datetime-diff-check-args
+  "Raises an exception if the database-type of `x` and `y` do not match the given predicate.
+   Note this doesn't raise an error if the database-type is nil, which can be the case for some drivers."
+  [x y pred]
+  (doseq [arg [x y]
+          :let [db-type (hx/database-type arg)]
+          :when (and db-type (not (pred db-type)))]
+    (throw (ex-info (tru "datetimeDiff only allows datetime, timestamp, or date types. Found {0}"
+                         (pr-str db-type))
+                    {:found db-type
+                     :type  qp.error-type/invalid-query}))))
+
+(defmethod ->honeysql [:sql :datetime-diff]
+  [driver [_ x y unit]]
+  (let [x (->honeysql driver x)
+        y (->honeysql driver y)]
+    (datetime-diff-check-args x y #(hx/is-of-type? % #"(?i)^(timestamp|date)"))
+    (datetime-diff driver unit x y)))
```

Comment:
I _think_ this regex can cover the types of all SQL implementations. If not, an individual implementation can modify it. 

Ideally I would have used a function like `hx/is-of-base-type` that can determine whether a field or value is of a certain base type. But sadly that's a little tricky to implement because type-info can be normalized, but `sql-jdbc.sync/database-type->base-type` expects a non-normalized database-type.